### PR TITLE
Feature/services

### PR DESCRIPTION
**Pull request by JNdhlovu**
Story: https://app.shortcut.com/smileid/story/9436/use-services-endpoint-to-fetch-validation-data

## Summary

Add the ability to query the smile services endpoint in the SDK

## Known Issues

Tests not yet written, will be in the next PR

### CODE REVIEW

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    case year
+    case bankCode = "bank_code"
+    case citizenship
+    case country
+    case idType = "id_type"
+    case userId = "user_id"
+    case jobId = "job_id"
+    case unknown = "Unknown"
+
+    private enum CodingKeys: String, CodingKey {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:88]: Is CodingKeys necessary for enums?

File: Sources/SmileID/Classes/Networking/ServiceRunnable.swift [hunk 2]
```diff
@@ -12,33 +12,44 @@ protocol ServiceRunnable {
...
-        .eraseToAnyPublisher()
+                                 body: body)
+            .flatMap(serviceClient.send)
+            .eraseToAnyPublisher()
+    }
+
+    func get<U: Decodable>(to path: PathType) -> AnyPublisher<U, Error> {
+        return createRestRequest(path: path,
+                                 method: .get,
+                                 headers: [.contentType(value: "application/json")])
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/ServiceRunnable.swift:50]: GET requests don't have a request body, therefore the content type header should not be set

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        case enhancedKyc = "enhanced_kyc"
+        case docVerification = "doc_verification"
+        case enhancedKycSmartSelfie = "ekyc_smartselfie"
+    }
+}
+
+// MARK: - CountryInfo
+
+public struct CountryInfo: Codable {
+    let countryCode: String
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:42]: countryCode isn't actually returned by the api in this object. It was added on Android as a convenience, and has to be manually set.

1. Can you add a comment indicating this
2. How will the deserializer handle this?/How will this value actually get set?

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    private enum CodingKeys: String, CodingKey {
+        case countryCode
+        case name
+        case availableIdTypes = "id_types"
+    }
+}
+
+// MARK: - AvailableIdType
+
+public struct AvailableIdType: Codable {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:55]: Same comments for `CountryInfo` apply here for `idTypeKey`

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+public struct HostedWeb: Codable {
+    var basicKyc: CountryCodeToCountryInfo
+    var biometricKyc: CountryCodeToCountryInfo
+    var enhancedKyc: CountryCodeToCountryInfo
+    var docVerification: CountryCodeToCountryInfo
+    var enhancedKycSmartSelfie: CountryCodeToCountryInfo
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        basicKyc = try container.decode(CountryCodeToCountryInfo.self, forKey: .basicKyc)
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:32]: I'm slightly confused why we need to manually decode all these. Shouldn't declaring the CodingKeys take care of it? i.e. we are not doing this in other Models

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        case docVerification = "doc_verification"
+        case enhancedKycSmartSelfie = "ekyc_smartselfie"
+    }
+}
+
+// MARK: - CountryInfo
+/**
+ * The countryCode  field is not populated/returned by the API response, hence it being marked as
+ * mutable. However, it should be populated before usage of this class.when the response gets decoded
+ *the same applies to availableIdTypes
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:58]: nit: spacing/punctuation

```suggestion
 * The countryCode field is not populated/returned by the API response, hence it being marked as
 * mutable. However, it should be populated before usage of this class/when the response gets decoded.
 * The same applies to availableIdTypes
```

File: Sources/SmileID/Classes/Networking/Models/Services.swift [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        name = try container.decode(String.self, forKey: .name)
+        availableIdTypes = try container.decode(IdTypeKeyToAvailableIdType.self, forKey: .availableIdTypes)
+        availableIdTypes = availableIdTypes.toAvailableIdTypes()
+    }
+}
+
+// MARK: - AvailableIdType
+/**
+ * The  idTypeKey  field is not populated/returned by the API response, hence it being marked as
+ * mutable. However, it should be populated before usage of this class.when the response gets decoded
```
Comment (vanshg) [inline Sources/SmileID/Classes/Networking/Models/Services.swift:81]: same here

```suggestion
 * The  idTypeKey field is not populated/returned by the API response, hence it being marked as
 * mutable. However, it should be populated before usage of this class/when the response gets decoded
```

File: Sources/SmileID/Classes/SmileID.swift [hunk 8]
```diff
@@ -26,6 +26,8 @@ public class SmileID {
     internal static var config: Config!
     internal static var useSandbox = true
     public private(set) static var theme: SmileIdTheme = DefaultTheme()
+    @ObservedObject
+    internal static var navigationState = NavigationViewModel()
 
     public class func initialize(config: Config = try! Config(url: Bundle.main.url(forResource: "smile_config",
```
[the remark this answers is not in the document]
Comment (vanshg) [reply Sources/SmileID/Classes/SmileID.swift:32]: Is it possible to get Danger to ignore this "error" for this line? We _want_ to do a force try here

[the remark this answers is not in the document]
Comment (JubrilO) [reply Sources/SmileID/Classes/SmileID.swift:32]: @vanshg @JNdhlovu we already have this on the main branch. But it seems develop is out of sync with main. Made a PR to merge develop into main so we can delete the develop branch after.

**Review by vanshg — APPROVED** [review]

**Pull request merged**